### PR TITLE
feat: Generate table of contents code action

### PR DESCRIPTION
**Pull request by keynmol**
Very basic demonstration of how a "Create TOC" can be implemented.
Note that I don't know F# at all :D 

TODO:

1. Learn F#
2. ~Detect an existing TOC and update it when the code action is invoked (can even put diagnostics on the exact locations saying it's out of date)~
    
    ~This is actually non-trivial if done in a non-invasive manner - only heuristic that can possibly exist is "first block of nothing but links in the document" and it can be easily broken. Can also match those links to a set of headings for some overlap number.~

### CODE REVIEW

File: Marksman/Marksman.fsproj [hunk 1]
```diff
@@ -1,39 +1,34 @@
...
-        <PackageReference Include="Serilog.Sinks.Console" Version="4.0.1" />
-    </ItemGroup>
-
-    <ItemGroup>
-        <ProjectReference Include="..\LanguageServerProtocol\LanguageServerProtocol.fsproj" />
-        <ProjectReference Include="..\MarkdigPatches\MarkdigPatches.csproj" />
-    </ItemGroup>
-
-</Project>
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (artempyanykh) [inline Marksman/Marksman.fsproj:1]: Is it some tool that reformat this file?

File: Marksman/Server.fs [hunk 2]
```diff
@@ -81,6 +81,11 @@ let mkServerCaps (par: InitializeParams) : ServerCapabilities =
 
     let clientDesc = ClientDescription.ofParams par
 
+    let codeActionOptions = {
```
Comment (artempyanykh) [inline Marksman/Server.fs:84]: `dotnet restore` and then `make fmt` ;)

File: Marksman/Server.fs [hunk 3]
```diff
@@ -715,4 +722,53 @@ type MarksmanServer(client: MarksmanClient) =
...
         AsyncLspResult.success tokens
 
+    override this.TextDocumentCodeAction(opts: CodeActionParams) =
+        let state = requireState ()
+        let docPath = opts.TextDocument.Uri |> PathUri.fromString
+        let range = opts.Range
+
+        let p = State.tryFindDocument docPath state
+        
+        let toc = Array.ofSeq (Cst.elementsAll p.Value.cst) |>
```
Comment (artempyanykh) [inline Marksman/Server.fs:732]: Maybe `doc.index |> Index.headings`?

Comment (keynmol) [reply Marksman/Server.fs:732]: wow there cowboy, slowdown with your academic magic

File: Marksman/Server.fs (hunk 2 above)
Comment (keynmol) [reply Marksman/Server.fs:84]: Apparently it's `dotnet tool restore`

File: Marksman/Server.fs [hunk 4]
```diff
@@ -715,4 +722,53 @@ type MarksmanServer(client: MarksmanClient) =
 
         AsyncLspResult.success tokens
 
+    override this.TextDocumentCodeAction(opts: CodeActionParams) =
+        let state = requireState ()
+        let docPath = opts.TextDocument.Uri |> PathUri.fromString
+        let range = opts.Range
+
+        let p = State.tryFindDocument docPath state
```
Comment (artempyanykh) [inline Marksman/Server.fs:730]: Perhaps match on the option?
```
let action = 
  match State.tryFindDocument docPath state with
  | None  ...
  | Some ...
```

Or you could use a `monad { }` [computation expression](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/computation-expressions) which is like for comprehension or do notation:
```
let action =
  monad {
    let! doc = State.tryFindDoc docPath state
    let toc = Toc.mk doc.index
  }
```

File: Marksman/Server.fs [hunk 5]
```diff
@@ -715,4 +722,53 @@ type MarksmanServer(client: MarksmanClient) =
...
+            Edit = Some {
+                Changes = Some mp;
+                DocumentChanges = None;
+            }
+        }
+
+        let commands = 
+            TextDocumentCodeActionResult.CodeActions [| codeAction |]
+            
+        async.Return (LspResult.Ok (Some commands))
```
Comment (artempyanykh) [inline Marksman/Server.fs:768]: AsyncLspResult.success (...)

**Comment by artempyanykh** [comment]
> Detect an existing TOC and update it when the code action is invoked (can even put diagnostics on the exact locations saying it's out of date)

> This is actually non-trivial if done in a non-invasive manner - only heuristic that can possibly exist is "first block of nothing but links in the document" and it can be easily broken. Can also match those links to a set of headings for some overlap number.

@keynmol what do you think about generating toc with a heading `## Table of Contents` (or with a slug of the heading being `table-of-contents` and then consider this heading and everything under it as a toc?

**Comment by artempyanykh** [comment]
Cool stuff!
giphy

File: Marksman/Index.fs [hunk 6]
```diff
@@ -16,8 +17,9 @@ type Index =
 module Index =
     let ofCst (cst: Cst) : Index =
         let titles = ResizeArray()
-        let headings = Dictionary<Slug, ResizeArray<Node<Heading>>>()
+        let headingsBySlug = Dictionary<Slug, ResizeArray<Node<Heading>>>()
         let wikiLinks = ResizeArray()
+        let headings = ResizeArray()
```
Comment (keynmol) [inline Marksman/Index.fs:22]: We need headings specifically in order (not just titles)

File: Marksman/Marksman.fsproj (hunk 1 above)
Comment (keynmol) [reply Marksman/Marksman.fsproj:1]: :shrug: no idea

_could be_ line endings? Are developing this on a Windows?

File: Marksman/Toc.fs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+open Marksman.Misc
+open Marksman.Index
+open Marksman.Cst
+open Marksman.Text
+open Ionide.LanguageServerProtocol.Types
+open Ionide.LanguageServerProtocol.Logging
+
+[<Literal>]
+let Marker = """<!--toc-->"""
```
Comment (keynmol) [inline Marksman/Toc.fs:11]: Can be configurable in `.marksman.toml` but I haven't looked into how to do that or whether it's useful

**Comment by keynmol** [comment]
One issue with this is that currently Marksman's parser detects YAML frontmatter stuff like:

```
---
title: Hello
---
```

As a heading, and adds it to the TOC. I believe this is parser's deficiency, as it interprets the yaml block (which arguably **isn't** a part of the markdown spec) as one of the syntaxes for headings:

image

What makes matters worse is that `---` is a horizontal line (https://spec.commonmark.org/0.30/#example-43), so you can't even technically parse the entire yaml frontmatter away :) 

# Markdown is hard

Only heuristic is really that the first line of the file is `---` and there's a closing one as well, and the user somehow indicated that they want yaml frontmatter detection enabled

File: Marksman/Marksman.fsproj (hunk 1 above)
Comment (artempyanykh) [reply Marksman/Marksman.fsproj:1]: I don't think it's line endings, bc the indentation here is 2 spaces vs 4 before. Could you rebase on latest master and run `make fmt` -- it should reformat the proj file.

File: Marksman/Server.fs [hunk 8]
```diff
@@ -715,4 +721,57 @@ type MarksmanServer(client: MarksmanClient) =
 
         AsyncLspResult.success tokens
 
+    override this.TextDocumentCodeAction(opts: CodeActionParams) =
+        let state = requireState ()
+        let docPath = opts.TextDocument.Uri |> PathUri.fromString
+
+        let documentBeginning =
+            { Start = { Line = 0; Character = 0 }
```
Comment (artempyanykh) [inline Marksman/Server.fs:729]: nit: `open Marksman.Misc` (if not already) to load `Position` and `Range` extensions and then `Position.Mk(0, 0)` or `Range.Mk(0, 0, 0, 0)`.

File: Marksman/Server.fs [hunk 9]
```diff
@@ -715,4 +721,57 @@ type MarksmanServer(client: MarksmanClient) =
...
+              End = { Line = 0; Character = 0 } }
+
+        let editAt rangeOpt text =
+            let textEdit =
+                { NewText = text
+                  Range = Option.defaultValue documentBeginning rangeOpt }
+
+            let mp = Map.empty.Add(opts.TextDocument.Uri, [| textEdit |])
```
Comment (artempyanykh) [inline Marksman/Server.fs:737]: nit: `Map.ofList [opts.TextDocument.Uri, [| textEdit |]]

File: Marksman/Server.fs [hunk 10]
```diff
@@ -715,4 +721,57 @@ type MarksmanServer(client: MarksmanClient) =
...
+              Edit = Some edit }
+
+        let codeActions =
+            match toc with
+            | None -> Array.empty
+            | Some (render, existing) ->
+                match existing with
+                | None -> [| codeAction "Create a Table of Contents" (editAt None render) |]
+                | Some (oldTocRange) ->
+                    [| codeAction "Update Table of Contents" (editAt (Some oldTocRange) render) |]
```
Comment (artempyanykh) [inline Marksman/Server.fs:771]: nit: *a* Table of Contents?

File: Marksman/Toc.fs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    let renderLink entry =
+        let offset = String.replicate (entry.Level - 1) " "
+        let slug = entry.Link |> Slug.toString
+        $"{offset}- [{entry.Title}](#{slug})"
+
+    let fromHeading (heading: Heading) : Entry =
+        let slug = Heading.slug heading
+        { Level = heading.level; Link = slug; Title = heading.title.text }
+
+type TableOfContents = { Entries: Entry[] }
```
Comment (artempyanykh) [inline Marksman/Toc.fs:28]: iirc formatter formats postfix `[]` array type weirdly. Better use `array<Entry>` instead

File: Marksman/Server.fs (hunk 10 above)
Comment (keynmol) [reply Marksman/Server.fs:771]: I think in this case "the" will be more grammatically correct :)

File: Marksman/Toc.fs [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    let mk (index: Index) : TableOfContents option =
+        let headings = index.headings |> Array.map (fun x -> x.data)
+
+        if index.headings.Length.Equals 0 then
+            None
+        else
+            Some { Entries = Array.map Entry.fromHeading headings }
+
+    let render (toc: TableOfContents) =
+        Marker
```
Comment (artempyanykh) [inline Marksman/Toc.fs:42]: nit: maybe `String.concat Environment.newLines [your-lines]`?

File: Marksman/Toc.fs [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+                            Collecting(extraLine, 2)
+                    else
+                        go (i + 1) (Collecting(expandToThisLine range, 0))
+                | BeforeMarker -> go (i + 1) BeforeMarker
+
+        match (go 0 BeforeMarker) with
+        // marker was never found
+        | Collecting (range, 2) -> Some range
+        | other ->
+            logger.error (Log.setMessage $"Got {other} instead")
```
Comment (artempyanykh) [inline Marksman/Toc.fs:93]: maybe warn?

also you could do `Log.setMessage "[what went wrong]" >> Log.addContext "state" other` -- structured logging FTW

File: Tests/Helpers.fs [hunk 14]
```diff
@@ -18,3 +19,11 @@ let checkInlineSnapshot (fmt: 'a -> string) (things: seq<'a>) (snapshot: seq<str
     let lines = Seq.map (fun x -> (fmt x).Lines()) things |> Seq.concat
 
     lines.ShouldMatchInlineSnapshot(snapshot)
+
+let makeFakeDocument (content: string) : Doc =
+    let text = Text.mkText content
+    Doc.mk (PathUri.fromString "memory://fake.md") (PathUri.fromString "memory://") None text
+
+let makeFakeDocumentLines (lines: array<string>) : Doc =
+    let text = Text.mkText (String.concat "\n" lines)
```
Comment (artempyanykh) [inline Tests/Helpers.fs:28]: "\n" -> Environment.NewLine ?

File: Tests/TocTests.fs [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+            TableOfContents.mk doc.index |> Option.get |> TableOfContents.render
+
+        let expectedLines =
+            [| Toc.Marker
+               "- [T1](#t1)"
+               " - [T2](#t2)"
+               "  - [T3](#t3)"
+               " - [T4](#t4)"
+               "  - [T5](#t5)"
+               "" // two new lines are important
```
Comment (artempyanykh) [inline Tests/TocTests.fs:83]: so, two lines are an end marker for toc? any reason you don't want ?
```
<!-- toc: start -->
<!-- toc: end -->
```

File: Marksman/Server.fs (hunk 10 above)
Comment (artempyanykh) [reply Marksman/Server.fs:771]: I trust you ;)

**Comment by artempyanykh** [comment]
> One issue with this is that currently Marksman's parser detects YAML frontmatter stuff like:

@keynmol I think `markdig` (the lib we use) has some support for yaml front matter -- you may want to look into mark dig pipeline in `Parser.fs`, perhaps there's some extension that needs to be activated to fix front matter handling.

**Comment by artempyanykh** [comment]
Awesome! 🎉 Thanks for the contribution ❤️

**Pull request merged**